core/tests/unit/ldml/ldml_test_source.cpp

```diff
@@ -482,6 +481,164 @@ int LdmlJsonTestSource::load(const nlohmann::json &data) {
+  std::u16string chstr = std::u16string(ch16.ch, len);
+  // append to expected
+  expected.append(chstr);
+  need_check = true;
+
+  // ---------------------------------------------------
+  // find a key that can emit this string.
+  // TODO-LDML: no transforms yet.
+  // TODO-LDML: looking for an exact single key for now
+
+  assert(kmxplus != nullptr);
+  // lookup the id
+  assert(kmxplus->strs != nullptr);
+  assert(kmxplus->key2 != nullptr);
+  assert(kmxplus->layr != nullptr);
+
+  assert(kmxplus->key2Helper.valid());
+  assert(kmxplus->layrHelper.valid());
+
+  // First, find the string as an id
+  // TODO-LDML: will not work for multi string cases
+  KMX_DWORD strId = kmxplus->strs->find(chstr);
+  if (strId == 0) { // will also get here if id is empty.
+    fillin.string = u"No string for repertoire test: ";
+    fillin.string.append(chstr);
+    fillin.type = LDML_ACTION_FAIL;
+    return;
+  }
+  assert(strId != 0);
+
+  // OK. Now we can search the keybag
+  KMX_DWORD keyIndex = 0;
+  auto *key2 = kmxplus->key2Helper.findKeyByStringTo(strId, keyIndex);
+  if (key2 == nullptr) {
+    fillin.string = u"No key for repertoire test: ";
+    fillin.string.append(chstr);
+    fillin.type = LDML_ACTION_FAIL;
+    return;
+  }
+
+  // Now, look for the _first_ candidate vkey match in the kmap.
+  for (KMX_DWORD kmapIndex = 0; kmapIndex < kmxplus->key2->kmapCount; kmapIndex++) {
+    auto *kmap = kmxplus->key2Helper.getKmap(kmapIndex);
+    assert(kmap != nullptr);
+    if (kmap->key == keyIndex) {
+      fillin.k = {(km_kbp_virtual_key)kmap->vkey, (uint16_t)kmap->mod};
+      std::cout << "found vkey " << fillin.k.vk << ":" << fillin.k.modifier_state << std::endl;
+      fillin.type = LDML_ACTION_KEY_EVENT;
+      return;
+    }
+  }
+
+  fillin.type = LDML_ACTION_FAIL;
+  fillin.string = u"Could not find candidate vkey: ";
+  fillin.string.append(chstr);
+}
+
+const std::u16string &
+LdmlJsonRepertoireTestSource::get_context() const {
+  return context; // no context needed
+}
+
+int LdmlJsonRepertoireTestSource::load(const nlohmann::json &data) {
+  this->data = data;  // TODO-LDML
+  // Need an update to json.hpp to use contains()
+  // if (data.contains("/type"_json_pointer)) {
+  if (data.find("type") != data.end()) {
+    type = data["/type"_json_pointer].get<std::string>();
+  } else {
+    type = "default";
+  }
+  chars      = data["/chars"_json_pointer].get<std::string>();
+  std::cout << "Loaded " << path << " = " << this->type << " || " << this->chars << std::endl;
+  UErrorCode status = U_ZERO_ERROR;
+  icu::UnicodeString pattern(chars.data(), chars.length());
+  uset = std::unique_ptr<icu::UnicodeSet>(new icu::UnicodeSet(pattern, status));
+  if (U_FAILURE(status)) {
+    // could be a malformed syntax isue
+    std::cerr << "UnicodeSet c'tor problem " << u_errorName(status) << std::endl;
+    return 1;
+  }
+  std::cout << "Got UnicodeSet of " << uset->size() << " char(s)." << std::endl;
+  #if (U_ICU_VERSION_MAJOR_NUM >= 70)
```

Comment:
> > What would you think of falling back to `pkg-config` (i.e. including existing icu) if meson was too old? Or some kind of build switch. Then Windows, and anyone else with later meson, can download icu via the wrap, and debian can include the icu4c package which is already available there.
> 
> Possible? @ermshiperete can you weigh in?

see https://github.com/keymanapp/keyman/pull/8441/commits/0693e6436df27ceb8544df5bb7f4283ac353f420#diff-cb5653763d2d663a6a5174df34f9fe7fa49d0f2d6f522d6f468309b404220726R7-R8

it would work like that, behind a conditional.

Reply:
the v70 has to do with the hasStrings function. But the general issue is dependent on where the spec finally lands. Like to leave this as a TODO-LDML for this clause.

Reply:
There's another way to rewrite it without hasStrings that i'll come back to.

Reply:
Ubuntu 20.04 Focal comes with ICU 66, so  things would be easier if we can make that the minimum instead of v70.

Reply:
ICU 66 is fine. This doesn't error, just has a different code path for <70